### PR TITLE
Enumerate available options in IPA installer

### PR DESCRIPTION
**Pull request by Akasurde**
Fix adds enumerated list of available options in IPA server
installer and IPA CA installer help options

Fixes https://fedorahosted.org/freeipa/ticket/5435

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

### CODE REVIEW

**Comment by Akasurde** [comment]
ping

**Comment by mbasti-rh** [comment]
Hi, we changed a lot of code during refactoring, PR doesn't apply. IMO list of choices shown in --help should be handled in knob() if metavar is not specified and type is choice

**Comment by jcholast** [comment]
@mbasti-rh: `knob()` already handles choices, it's the built-in `optparse` module which does not display them. Once the installer code is migrated to `argparse`, this problem will go away.

**Comment by mbasti-rh** [comment]
@jcholast I know, but it doesn't fill `metavar` with choices. I don't know when we will migrate to argparse, so I think until that we can extend it to show choices with optparse too

**Comment by mbasti-rh** [comment]
@jcholast any update? Should reject this PR and wait for `argparse` or fix it with `optparse` as well?  IMO fixing it now is better for UX, we dont know when or if we migrate to argparse.

**Comment by jcholast** [comment]
@mbasti-rh, I don't care as long as it's done right (i.e. without hardcoding `cli_metavar` in knob definitions).

**Comment by Akasurde** [comment]
@jcholast @mbasti-rh I will work on modifying `Knob()` to handle metavar

**Comment by jcholast** [comment]
@Akasurde, `Knob()` already handles metavar properly, you need to work on the interface between the installer and `optparse` - `ipapython.install.cli`.

File: ipapython/install/cli.py [hunk 1]
```diff
@@ -179,6 +179,7 @@ def add_options(cls, parser, positional=False):
             elif issubclass(knob_scalar_type, enum.Enum):
                 kwargs['type'] = 'choice'
                 kwargs['choices'] = [i.value for i in knob_scalar_type]
+                kwargs['metavar'] = "[%s]" % ("|".join(kwargs['choices']))
```
Comment (HonzaCholasta) [inline ipapython/install/cli.py]: - Please use the same format as `argparse`: `{a,b,c}`.
- [`str.format()`](https://docs.python.org/3/library/stdtypes.html#str.format) is the preferred way of formatting strings.

File: ipapython/install/cli.py [hunk 2]
```diff
@@ -179,6 +179,8 @@ def add_options(cls, parser, positional=False):
             elif issubclass(knob_scalar_type, enum.Enum):
                 kwargs['type'] = 'choice'
                 kwargs['choices'] = [i.value for i in knob_scalar_type]
+                kwargs['metavar'] = "{{{0}}}".format(
+                                                "|".join(kwargs['choices']))
```
Comment (HonzaCholasta) [inline ipapython/install/cli.py]: Please use `,` rather than `|`, as `argparse` does.

**Comment by HonzaCholasta** [comment]
Works for me, although you should probably keep the changes to `ipa-ca-install` from the original patch (using the `argparse` format, of course).

File: install/tools/ipa-ca-install [hunk 3]
```diff
@@ -61,15 +61,18 @@ def parse_options():
...
+                      metavar="{{{0}}}".format(",".join(ext_cas)),
                       help="Type of the external CA")
     parser.add_option("--external-cert-file", dest="external_cert_files",
                       action="append", metavar="FILE",
                       help="File containing the IPA CA certificate and the external CA certificate chain")
+    ca_algos = ('SHA1withRSA', 'SHA256withRSA', 'SHA512withRSA')
     parser.add_option("--ca-signing-algorithm", dest="ca_signing_algorithm",
-                      type="choice",
-                      choices=('SHA1withRSA', 'SHA256withRSA', 'SHA512withRSA'),
+                      type="choice", choices=ca_alogs,
```
Comment (HonzaCholasta) [inline install/tools/ipa-ca-install]: Typo: `ca_alogs`.

**Comment by HonzaCholasta** [comment]
Fixed upstream
master:
https://fedorahosted.org/freeipa/changeset/80c0e5cb8d689cf1ec6a883d2c7000f9dadbf7d8

**Comment by Akasurde** [comment]
@HonzaCholasta Thanks for your help and reviews.

**Pull request closed without merging**